core/services/headreporter/head_reporter.go

```diff
@@ -0,0 +1,129 @@
+
+	evmtypes "github.com/smartcontractkit/chainlink/v2/core/chains/evm/types"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/legacyevm"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+)
+
+//go:generate mockery --quiet --name HeadReporter --output ../../internal/mocks/ --case=underscore
+type (
+	HeadReporter interface {
+		ReportNewHead(ctx context.Context, head *evmtypes.Head) error
+		ReportPeriodic(ctx context.Context) error
+	}
+
+	HeadReporterService struct {
+		services.StateMachine
+		ds           sqlutil.DataSource
+		chains       legacyevm.LegacyChainContainer
+		lggr         logger.Logger
+		newHeads     *mailbox.Mailbox[*evmtypes.Head]
+		chStop       services.StopChan
+		wgDone       sync.WaitGroup
+		reportPeriod time.Duration
+		reporters    []HeadReporter
+	}
+)
+
+func NewHeadReporterService(config config.HeadReport, ds sqlutil.DataSource, chainContainer legacyevm.LegacyChainContainer, lggr logger.Logger, monitoringEndpointGen telemetry.MonitoringEndpointGenerator, opts ...interface{}) *HeadReporterService {
+	reporters := make([]HeadReporter, 2)
+	reporters = append(reporters, NewPrometheusReporter(ds, chainContainer, opts))
+	if config.TelemetryEnabled() {
+		reporters = append(reporters, NewTelemetryReporter(chainContainer, lggr, monitoringEndpointGen))
+	}
+	return NewHeadReporterServiceWithReporters(ds, chainContainer, lggr, reporters, opts)
+}
+
+func NewHeadReporterServiceWithReporters(ds sqlutil.DataSource, chainContainer legacyevm.LegacyChainContainer, lggr logger.Logger, reporters []HeadReporter, opts ...interface{}) *HeadReporterService {
+	reportPeriod := 30 * time.Second
+	for _, opt := range opts {
+		switch v := opt.(type) {
+		case time.Duration:
+			reportPeriod = v
+		}
+	}
+	chStop := make(chan struct{})
+	return &HeadReporterService{
+		ds:           ds,
+		chains:       chainContainer,
+		lggr:         lggr.Named("HeadReporterService"),
+		newHeads:     mailbox.NewSingle[*evmtypes.Head](),
+		chStop:       chStop,
+		wgDone:       sync.WaitGroup{},
+		reportPeriod: reportPeriod,
+		reporters:    reporters,
+	}
+}
+
+func (hrd *HeadReporterService) Start(context.Context) error {
+	return hrd.StartOnce(hrd.Name(), func() error {
+		hrd.wgDone.Add(1)
+		go hrd.eventLoop()
+		return nil
+	})
+}
+
+func (hrd *HeadReporterService) Close() error {
+	return hrd.StopOnce(hrd.Name(), func() error {
+		close(hrd.chStop)
+		hrd.wgDone.Wait()
+		return nil
+	})
+}
+
+func (hrd *HeadReporterService) Name() string {
+	return hrd.lggr.Name()
+}
+
+func (hrd *HeadReporterService) HealthReport() map[string]error {
+	return map[string]error{hrd.Name(): hrd.Healthy()}
+}
+
+func (hrd *HeadReporterService) OnNewLongestChain(ctx context.Context, head *evmtypes.Head) {
+	hrd.newHeads.Deliver(head)
+}
+
+func (hrd *HeadReporterService) eventLoop() {
+	hrd.lggr.Debug("Starting event loop")
+	defer hrd.wgDone.Done()
+	ctx, cancel := hrd.chStop.NewCtx()
+	defer cancel()
+	after := time.After(hrd.reportPeriod)
```

Comment:
Just a heads-up: this is fixing a bug that was present in the original code, where periodically reported Prom metrics were inadvertently being set only if it went through (default=) 30s without a new head being detected;

Since the metrics being set periodically are different than the ones on new heads, this means those could go a long time (or even forever) without being updated.

Taking `after` out of the `for` loop ensures these are set unconditionally every `reportPeriod`, as expected